File: ios/Runner/PeerManager.swift
//
//  PeerManager.swift
//  Runner
//
//  Created by Sudharshan on 3/29/19.
//  Copyright © 2019 The Chromium Authors. All rights reserved.
//

import Foundation
import WebRTC
import Just
import Starscream
import PercentEncoder

public enum RTCWrapperState {
    case disconnected
    case connected
    case connecting
}

class PeerManager: NSObject {
    //****************************************************************************************************
    // GLOBAL CLASS ATTRIBUTES
    //****************************************************************************************************
    
    // State
    var state: RTCWrapperState = .disconnected
    var activeConversation: String?
    var authToken: String?
    
    // Required interfaces
    var connectionFactory: RTCPeerConnectionFactory?
    var socket: WebSocket?
    
    // WebRTC
    var peerConnection: RTCPeerConnection?
    var remoteIceCandidates: [RTCIceCandidate] = []
    
    // Configuration
    let iceServers: [RTCIceServer] = [RTCIceServer(urlStrings: ["stun:stun.l.google.com:19302"])]
    let connectionConstraint = RTCMediaConstraints(mandatoryConstraints: nil, optionalConstraints: ["DtlsSrtpKeyAgreement": "true"])
    let channelConstraint = RTCMediaConstraints(mandatoryConstraints: ["OfferToReceiveAudio" : "true"], optionalConstraints: nil)
    
    // Streams
    var localMediaStream: RTCMediaStream!
    var localAudioTrack: RTCAudioTrack!
    var remoteAudioTrack: RTCAudioTrack!
    
    //****************************************************************************************************
    // START OF PUB FUNCTIONS AND INIT
    //****************************************************************************************************
    
    public override init() {
        super.init()
        
        // Initialize peer factory
        RTCPeerConnectionFactory.initialize()
        self.connectionFactory = RTCPeerConnectionFactory()
    }
    
    // MUST CALL THIS BEFORE STUFF WORKS
    public func initializeToken(authToken: String) {
        self.authToken = authToken
    }
    
    public func join(conversationId: String) {
        self.activeConversation = conversationId
        
        // Setup socket and peerconnection if it doesn't exist
        if (self.socket == nil) {
            var request = URLRequest(url: URL(string: "ws://localhost:8080/")!)
            request.setValue("Bearer \(self.authToken ?? "0")", forHTTPHeaderField: "Authorization")
            
            self.socket = WebSocket(request: request)
            self.socket?.delegate = self
            self.socket?.connect()
        } else if(self.peerConnection == nil) {
            self.connect()
        }
        
        // Configure SFU conversation Id
        Just.post("http://localhost/join/\(conversationId)",
            headers: ["Authorization": "Bearer \(self.authToken ?? "0")"])
    }
    
    public func exit() {
        activeConversation = nil
        self.disconnect()
        self.socket?.disconnect()
    }
    
    public func get() -> String? {
        return activeConversation
    }
}

//****************************************************************************************************
// START OF PRIVATE FUNCTIONS
//****************************************************************************************************

private extension PeerManager {
    func connect() {
        // Create peerconnection
        let configuration = RTCConfiguration()
        configuration.iceServers = self.iceServers;
        let peerConnection = self.connectionFactory?.peerConnection(with: configuration, constraints: self.connectionConstraint, delegate: self)
        
        // Add localstream to peerconnection
        self.state = .connecting
        let localStream = self.localStream()
        peerConnection?.add(localStream)
        
        // Store peerconnection
        self.peerConnection = peerConnection
        self.createOffer()
    }
    
    func disconnect() {
        if let peerConnection = self.peerConnection {
            peerConnection.close()
            if let stream = peerConnection.localStreams.first {
                peerConnection.remove(stream)
            }
            
            // Reset state
            self.peerConnection = nil
            self.state = .disconnected
        }
    }
    
    func createOffer() {
        if let peerConnection = self.peerConnection {
            
            peerConnection.offer(for: self.channelConstraint, completionHandler: { [weak self]  (sdp, error) in
                // Exit if this object doesn't exist anymore cause it is a weak link
                guard let this = self else { return }
                
                if let error = error {
                    print(error)
                } else {
                    // Use the sdp generated
                    guard let sdp = sdp else{
                        return
                    }

                    this.peerConnection?.setLocalDescription(sdp, completionHandler: {[weak self] (error) in
                        guard let _ = self, let error = error else { return }
                        print(error)
                    })
                    
                    // Post it to the server
                    this.socket?.write(string: "offer::\(sdp.sdp)")
                }
            })
        }
    }
    
    func createAnswerForOffer(remoteSdp: String) {
        if let peerConnection = self.peerConnection {
            let sessionDescription = RTCSessionDescription.init(type: .offer, sdp: remoteSdp)
            peerConnection.setRemoteDescription(sessionDescription, completionHandler: { [weak self] (error) in
                // Exit if this object doesn't exist anymore cause it is a weak link
                guard let this = self else { return }
                
                if let error = error {
                    // Throw an error
                    print(error)
                } else {
                    // Add ice-candidates after setting remote description
                    for iceCandidate in this.remoteIceCandidates {
                        peerConnection.add(iceCandidate)
                    }
                    this.remoteIceCandidates = []
                    
                    // create answer
                    peerConnection.answer(for: this.channelConstraint, completionHandler:
                        { (sdp, error) in
                            if let error = error {
                                // Throw an error
                                print(error)
                            } else {
                                guard let sdp = sdp else{
                                    return
                                }
                                // add generated local sdp
                                peerConnection.setLocalDescription(sdp, completionHandler: {[weak self] (error) in
                                    guard let _ = self, let error = error else { return }
                                    print(error)
                                })
                                
                                // Send the localsdp to the server
                                this.socket?.write(string: "answer::\(sdp.sdp)")
                                this.state = .connected
                            }
                    })
                }
            })
        }
    }
    
    func handleAnswer(remoteSdp: String) {
        if let peerConnection = self.peerConnection {
            let sessionDescription = RTCSessionDescription.init(type: .answer, sdp: remoteSdp)
            
            peerConnection.setRemoteDescription(sessionDescription, completionHandler: { [weak self] (error) in
                // Exit if this object doesn't exist anymore cause it is a weak link
                guard let this = self else { return }
                
                if let error = error {
                    // Throw an error
                    print(error)
                } else {
                    // Add ice-candidates after setting remote description
                    for iceCandidate in this.remoteIceCandidates {
                        peerConnection.add(iceCandidate)
                    }
                    this.remoteIceCandidates = []
                    this.state = .connected
                }
            })
        }
    }

    
    func addIceCandidate(iceCandidate: RTCIceCandidate) {
        // Set ice candidate after setting remote description
        if self.peerConnection?.remoteDescription != nil {
            self.peerConnection?.add(iceCandidate)
        } else {
            self.remoteIceCandidates.append(iceCandidate)
        }
    }
    
    // Utility function for creating a localStream
    func localStream() -> RTCMediaStream {
        let factory = self.connectionFactory!
        let localStream = factory.mediaStream(withStreamId: "RTCmS")
        
        let audioTrack = factory.audioTrack(withTrackId: "RTCaS0")
        localStream.addAudioTrack(audioTrack)
        
        return localStream
    }
}

//****************************************************************************************************
// START OF WEBRTC DELEGATIONS
//****************************************************************************************************
extension PeerManager: RTCPeerConnectionDelegate {
    func peerConnectionShouldNegotiate(_ peerConnection: RTCPeerConnection) {
        
    }
    
    func peerConnection(_ peerConnection: RTCPeerConnection, didChange stateChanged: RTCSignalingState) {
        
    }
    
    func peerConnection(_ peerConnection: RTCPeerConnection, didAdd stream: RTCMediaStream) {
        print("adding new stream from remote")
        self.remoteAudioTrack = stream.audioTracks[0]
    }
    
    func peerConnection(_ peerConnection: RTCPeerConnection, didRemove stream: RTCMediaStream) {
        
    }
    
    func peerConnection(_ peerConnection: RTCPeerConnection, didChange newState: RTCIceConnectionState) {
        
    }
    
    func peerConnection(_ peerConnection: RTCPeerConnection, didChange newState: RTCIceGatheringState) {
        
    }
    
    func peerConnection(_ peerConnection: RTCPeerConnection, didGenerate candidate: RTCIceCandidate) {
        let candidateString = "\(candidate.sdp)::\(candidate.sdpMLineIndex)::\(candidate.sdpMid ?? "")"
        self.socket?.write(string: "ice::\(candidateString)")
    }
    
    func peerConnection(_ peerConnection: RTCPeerConnection, didRemove candidates: [RTCIceCandidate]) {
        
    }
    
    func peerConnection(_ peerConnection: RTCPeerConnection, didOpen dataChannel: RTCDataChannel) {
        
    }
}

//****************************************************************************************************
// START OF WEBSOCKET DELEGATIONS
//****************************************************************************************************
extension PeerManager: WebSocketDelegate {
    func websocketDidConnect(socket: WebSocketClient) {
        print("websocket is connected")
    }
    
    func websocketDidDisconnect(socket: WebSocketClient, error: Error?) {
        if let e = error as? WSError {
            print("websocket is disconnected: \(e.message)")
        } else if let e = error {
            print("websocket is disconnected: \(e.localizedDescription)")
        } else {
            print("websocket disconnected")
        }
    }
    
    func websocketDidReceiveData(socket: WebSocketClient, data: Data) {
        return
    }
    
    // FORMAT OF MESSAGE <message-type>::<message>
    // message-type: (offer, ice)
    // message:
    //  offer: <remoteSdp, String>
    //  ice: <sdp, String>::<sdpMLineIndex, String>::<sdpMid, String>
    
    func websocketDidReceiveMessage(socket: WebSocketClient, text: String) {
        let dataArr = text.components(separatedBy: "::")
        
        if(dataArr[0] == "offer") {
            // Handle offer
            
            // Check dataArr size of 2
            if dataArr.count != 2 {
                // Incorrect data format error
                return
            }
            
            self.createAnswerForOffer(remoteSdp: dataArr[1])
        } else if (dataArr[0] == "answer") {
            // Handle answer
            
            // Check dataArr size of 2
            if dataArr.count != 2 {
                // Incorrect data format error
                return
            }
            
            self.handleAnswer(remoteSdp: dataArr[1])
        } else if(dataArr[0] == "ice") {
            // Handle ice
            
            // Check dataArr size of 4
            if dataArr.count != 4 {
                // Incorrect data format error
                return
            }
            
            guard let sdpMLineIndex = Int32(dataArr[2]) else {
                // Invalid sdpMLineIndex error
                return
            }
            
            let iceCandidate = RTCIceCandidate(sdp: dataArr[1], sdpMLineIndex: sdpMLineIndex, sdpMid: dataArr[3])
            self.addIceCandidate(iceCandidate: iceCandidate)
        } else {
            // Invalid format error
            return
        }
    }
}

File: android/app/src/main/kotlin/com/example/frontend_flutter/PeerManager.kt
package com.example.frontend_flutter

import android.content.Context
import org.webrtc.PeerConnection.IceServer
import org.webrtc.PeerConnection.RTCConfiguration
import org.webrtc.PeerConnection.PeerConnectionState
import com.neovisionaries.ws.client.*
import org.webrtc.*
import java.net.URL
import org.webrtc.SessionDescription
import java.net.HttpURLConnection


class PeerManager(applicationContext: Context) {
    // State
    private var state: PeerConnectionState = PeerConnectionState.DISCONNECTED
    private var activeConversation: String? = null
    private var authToken: String? = null

    // Required interfaces
    private var connectionFactory: PeerConnectionFactory
    private var sockConnect: Boolean = false
    private var socket: WebSocket? = null

    // WebRTC
    private var peerConnection: PeerConnection? = null
    private var remoteIceCandidates = ArrayList<IceCandidate>()

    // Configuration
    private var iceServers: List<IceServer> = listOf(IceServer.builder("stun:stun.l.google.com:19302").createIceServer())
    private var channelConstraint = MediaConstraints().apply { mandatory.add(MediaConstraints.KeyValuePair("OfferToReceiveAudio", "true")) }

    // Streams
    private lateinit var remoteAudioTrack: AudioTrack


    //****************************************************************************************************
    // START OF PUB FUNCTIONS AND INIT
    //****************************************************************************************************

    init {
        PeerConnectionFactory.initialize(PeerConnectionFactory.InitializationOptions.builder(applicationContext).createInitializationOptions())
        connectionFactory = PeerConnectionFactory.builder().createPeerConnectionFactory()
    }

    // MUST CALL THIS BEFORE STUFF WORKS
    fun initializeToken(authToken: String) {
        this.authToken = authToken
    }

    fun join(conversationId: String) {
        activeConversation = conversationId

        // Setup socket and peerConnection if it doesn't exist
        if (socket == null) {
            socket = WebSocketFactory()
                    .createSocket(URL("wss://staging.beepvoice.app/webrtc/connect"))
                    .addHeader("Authorization", "Bearer ${authToken ?: "0"}")
                    .addHeader("Host", "staging.beepvoice.app")
                    .addListener(CustomWSAdapter())
                    .connect()


            // TODO: Test if disabling SSL cert verification is necessary
            //s.disableSSLCertValidation = true
        }

        // Configure SFU conversation Id
        URL("https://staging.beepvoice.app/webrtc/join/${conversationId}")
                .openConnection()
                .run {
                    // Cast to a HTTP connection
                    this as HttpURLConnection
                }.apply {
                    // Configure request and connect to the URL
                    requestMethod = "POST"
                    setRequestProperty("Authorization", "Bearer ${authToken ?: "0"}")
                    connect()
                }

        // Connect the PeerConnection
        connect()
    }

    fun exit() {
        activeConversation = null
        disconnect()
    }

    fun get(): String? {
        return activeConversation
    }


    //****************************************************************************************************
    // START OF PRIVATE FUNCTIONS
    //****************************************************************************************************
    private fun connect() {
        // Create peerConnection
        val configuration = RTCConfiguration(iceServers).apply { enableDtlsSrtp = true }
        peerConnection = connectionFactory.createPeerConnection(configuration, CustomPeerConnectionObserver())

        // Add localStream to peerConnection
        state = PeerConnectionState.CONNECTING
        peerConnection?.addStream(localStream())

        // Store peerConnection
        print("CONNECTING")
        createOffer()
    }

    private fun disconnect() {
        // Executes if peerConnection is not null
        peerConnection?.let {
            it.close()
            it.dispose()

            socket?.disconnect()
            socket?.clearListeners()

            while(sockConnect) {}

            peerConnection = null
            socket = null
            state = PeerConnectionState.DISCONNECTED
        }
    }

    private fun createOffer() {
        peerConnection?.createOffer(object : SdpObserver {
            override fun onCreateSuccess(sdp: SessionDescription) {
                peerConnection?.setLocalDescription(PrintSdpObserverError(), sdp)

                while (!sockConnect) {}

                socket?.sendText("offer::${sdp}")
            }

            override fun onCreateFailure(error: String) { print(error) }
            override fun onSetFailure(error: String) { print(error) }
            override fun onSetSuccess() {}
        }, channelConstraint)
    }

    private fun createAnswerForOffer(remoteSdp: String) {
        // Executes if peerConnection is not null
        peerConnection?.let {
            val sessionDescription = SessionDescription(SessionDescription.Type.OFFER, remoteSdp)

            it.setRemoteDescription(object: SdpObserver {
                override fun onSetSuccess() {
                    // Add ice-candidates after setting remote description
                    for (iceCandidate in remoteIceCandidates) {
                        it.addIceCandidate(iceCandidate)
                    }
                    remoteIceCandidates.clear()

                    // Create answer
                    it.createAnswer(object: SdpObserver {
                        override fun onCreateSuccess(sdp: SessionDescription) {
                            it.setLocalDescription(PrintSdpObserverError(), sdp)

                            // Send the localsdp to the server
                            socket?.sendText("answer::${sdp}")
                            state = PeerConnectionState.CONNECTED
                        }
                        override fun onCreateFailure(error: String) { print(error) }
                        override fun onSetFailure(error: String) { print(error) }
                        override fun onSetSuccess() {}
                    }, channelConstraint)
                }

                override fun onSetFailure(error: String) { print(error) }
                override fun onCreateFailure(error: String) { print(error) }
                override fun onCreateSuccess(sdp: SessionDescription) {}
            }, sessionDescription)
        }
    }

    private fun handleAnswer(remoteSdp: String) {
        // Executes if peerConnection is not null
        peerConnection?.let {
            val sessionDescription = SessionDescription(SessionDescription.Type.ANSWER, remoteSdp)

            it.setRemoteDescription(object: SdpObserver {
                override fun onSetSuccess() {
                    // Add ice-candidates after setting remote description
                    for (iceCandidate in remoteIceCandidates) {
                        it.addIceCandidate(iceCandidate)
                    }
                    remoteIceCandidates.clear()
                    state = PeerConnectionState.CONNECTED
                }

                override fun onCreateFailure(error: String) { print(error) }
                override fun onSetFailure(error: String) { print(error) }
                override fun onCreateSuccess(sdp: SessionDescription) {}
            }, sessionDescription)
        }
    }

    private fun addIceCandidate(iceCandidate: IceCandidate) {
        // Set ice candidate after setting remote description
        if (peerConnection?.remoteDescription != null) {
            peerConnection?.addIceCandidate(iceCandidate)
        } else {
            remoteIceCandidates.add(iceCandidate)
        }
    }

    // Utility function for creating a localStream
    private fun localStream(): MediaStream {
        val localStream = connectionFactory.createLocalMediaStream("RTCmS")
        val audioTrack = connectionFactory.createAudioTrack("RTCaS0", connectionFactory.createAudioSource(MediaConstraints()))
        localStream.addTrack(audioTrack)

        return localStream
    }


    //****************************************************************************************************
    // PeerConnection.Observer implementation
    //****************************************************************************************************
    inner class CustomPeerConnectionObserver: PeerConnection.Observer {
        override fun onSignalingChange(signalingState: PeerConnection.SignalingState) {}
        override fun onIceConnectionChange(iceConnectionState: PeerConnection.IceConnectionState) {}
        override fun onIceConnectionReceivingChange(b: Boolean) {}
        override fun onRemoveStream(mediaStream: MediaStream) {}
        override fun onDataChannel(dataChannel: DataChannel) {}
        override fun onRenegotiationNeeded() {}
        override fun onIceCandidatesRemoved(iceCandidates: Array<IceCandidate>) {}
        override fun onIceGatheringChange(iceGatheringState: PeerConnection.IceGatheringState) {}
        override fun onAddTrack(rtpReceiver: RtpReceiver, mediaStreams: Array<MediaStream>) {}

        override fun onIceCandidate(iceCandidate: IceCandidate) {
            // TODO: Check if this is necessary
            socket?.sendText("ice::${iceCandidate}")
        }

        override fun onAddStream(stream: MediaStream) {
            print("adding new stream from remote")
            remoteAudioTrack = stream.audioTracks[0]
        }
    }


    //****************************************************************************************************
    // Custom WebSocket listener for processing messages
    //****************************************************************************************************
    inner class CustomWSAdapter: WebSocketAdapter() {

        override fun onError(websocket: WebSocket?, cause: WebSocketException?) {
            // TODO: Check if cause.getError() produces better error messages
            print("WebSocket error: $cause")
        }

        override fun onConnected(websocket: WebSocket?, headers: MutableMap<String, MutableList<String>>?) {
            print("WebSocket is connected")
            sockConnect = true
        }

        override fun onDisconnected(websocket: WebSocket?, serverCloseFrame: WebSocketFrame?, clientCloseFrame: WebSocketFrame?, closedByServer: Boolean) {
            print("WebSocket is disconnected")
            sockConnect = false
        }

        // FORMAT OF MESSAGE <message-type>::<message>
        // message-type: (offer, ice)
        // message:
        //  offer: <remoteSdp, String>
        //  ice: <sdp, String>::<sdpMLineIndex, String>::<sdpMid, String>
        override fun onTextMessage(websocket: WebSocket?, text: String?) {
            val dataArr: List<String>? = text?.split("::")
            print(text)

            when (dataArr?.get(0)) {
                "offer" -> {
                    print("OFFER:\n${dataArr[1]}")
                    // Check dataArr size of 2
                    if (dataArr.count() != 2) {
                        // Incorrect data format error
                        return
                    }
                    this@PeerManager.createAnswerForOffer(dataArr[1])
                }
                "answer" -> {
                    // Handle answer
                    print("ANSWER:\n${dataArr[1]}")
                    // Check dataArr size of 2
                    if (dataArr.count() != 2) {
                        // Incorrect data format error
                        return
                    }
                    this@PeerManager.handleAnswer(dataArr[1])
                }
                "ice" -> {
                    // Handle ice
                    print("ICE:\n${dataArr[1]}")
                    // Check dataArr size of 2
                    if (dataArr.count() != 2) {
                        // Incorrect data format error
                        return
                    }
                    this@PeerManager.addIceCandidate(IceCandidate(null, 0, dataArr[1]))
                }
            }
        }
    }


    //****************************************************************************************************
    // Simple SdpObserver implementation that prints errors only
    //****************************************************************************************************
    class PrintSdpObserverError: SdpObserver {
        /** Called on error of Create{Offer,Answer}().  */
        override fun onCreateFailure(error: String) { print(error) }
        /** Called on error of Set{Local,Remote}Description().  */
        override fun onSetFailure(error: String) { print(error) }
        /** Called on success of Create{Offer,Answer}().  */
        override fun onCreateSuccess(sdp: SessionDescription) {}
        /** Called on success of Set{Local,Remote}Description().  */
        override fun onSetSuccess() {}
    }
}
